crates/bevy_transform/src/systems.rs

```diff
@@ -36,150 +44,159 @@ pub fn sync_simple_transforms(
-            //   - The caller ensures that each child has one and only one unique parent throughout the entire
-            //     hierarchy.
-            //
-            // For example, consider the following malformed hierarchy:
-            //
-            //     A
-            //   /   \
-            //  B     C
-            //   \   /
-            //     D
-            //
-            // D has two parents, B and C. If the propagation passes through C, but the ChildOf component on D points to B,
-            // the above check will panic as the origin parent does match the recorded parent.
-            //
-            // Also consider the following case, where A and B are roots:
-            //
-            //  A       B
-            //   \     /
-            //    C   D
-            //     \ /
-            //      E
-            //
-            // Even if these A and B start two separate tasks running in parallel, one of them will panic before attempting
-            // to mutably access E.
-            (unsafe { transform_query.get_unchecked(entity) }) else {
-                return;
-            };
+    let mut outbox = queue.local_queue.borrow_local_mut();
+    loop {
+        // Try to acquire a lock on the work queue in a tight loop.
+        let Ok(rx) = queue.receiver.try_lock() else {
+            continue;
+        };
+        // If the queue is empty and no other threads are busy processing work, we can conclude
+        // there is no more work to do, and end the task by exiting the loop.
+        let Some(mut tasks) = rx.try_iter().next() else {
+            if queue.busy_threads.load(Ordering::Relaxed) == 0 {
+                break; // All work is complete, kill the worker
+            }
+            continue; // No work to do now, but another thread is busy creating more work
+        };
+        if tasks.is_empty() {
+            continue;
+        }
 
-        changed |= transform.is_changed() || global_transform.is_added();
-        if changed {
-            *global_transform = parent.mul_transform(*transform);
+        // At this point, we know there is work to do, so we increment the busy thread counter,
+        // and drop the mutex guard *after* we have incremented the counter. This ensures that
+        // if another thread is able to acquire a lock, the busy thread counter will already be
+        // incremented.
+        queue.busy_threads.fetch_add(1, Ordering::Relaxed);
+        drop(rx);
+        for parent in tasks.drain(..) {
+            // SAFETY: Visiting the hierarchy as a tree without cycles
+            #[expect(unsafe_code, reason = "Mutating disjoint entities in parallel")]
+            unsafe {
+                let (_, _, parent_transform, children) = nodes.get_unchecked(parent).unwrap();
+                propagate_to_child_unchecked(parent_transform, children, nodes, &mut outbox);
+            }
+        }
+        for chunk in outbox.chunks(1024) {
+            queue.sender.send(chunk.to_vec()).ok();
```

Comment:
It should be slightly more efficient to send boxed slices instead of `Vec`s, would that work on the receiving end?
```suggestion
            queue.sender.send(Box::from(chunk)).ok();
```

Reply:
Just wonder if cloning the outbox to do 1 big allocation is better than multiple small allocations?

Reply:
> Is it because the Box is one usize but a Vec is 2 usize?
> 
> Or because `chunk.to_vec()` makes a copy of the vec, while `Box::from(chunk)` reuses the allocation?

`chunk` is a slice, both `.to_vec()` and `Box::from` will make an allocation, however what is being sent over the channel is just the pointer. `Box<[T]>` is 16 bytes, while `Vec<T>` is 24 bytes.
If these allocations are never extended, then the `capacity` field on `Vec` has no use here.

Reply:
I tried using a boxed slice, but profiling showed it was the same or slower than the vector.